# Linkfire candidate’s assessment – Product Manager’s role – Petr Agulin 

### Exercise 2

<div class="alert alert-block alert-danger">
    
Problem setting
    

Attached is data from 5 of our partners over a 2 week time frame showing visits (whenever the page was loaded) and ad impressions (whenever an ad was shown to a user). 
    

* What do you notice?
* Can you make statements about the partners, in comparison to each other?
* What would you investigate next?
    

Dataset:
https://docs.google.com/spreadsheets/d/1juE6ejW5y5oCR2Q6fmfuvw7vLOXuDTKZs8txWGqELSY/edit?usp=sharing_eil_se_dm&ts=62fa8e0c 

</div>


<div class="alert alert-block alert-success">
    
Solution
    

I have decided to analyse the dataset using my knowledge of Python. 
    
    
Below is my analysis of the dataset with comments of what I do and final conclusions.
</div>

<a id='Contents'></a>

### Contents

[1. Data upload and review](#01)

* [1.1 Import of Python libraries and original data](#02)
* [1.2 Review of the original table](#03)

[2. Data pre-processing and optimization](#04)

* [2.1 Check for full duplicates](#05)
* [2.2 Changing data formats for ease of use](#06)

[3. Exploratory data analysis](#07)

* [3.1 Analysis of ad impressions](#08)
* [3.2 Partner analysis](#09)

[4. Conclusions](#10)

<a id='01'></a>
### 1. Data upload and review

<a id='02'></a>
#### 1.1. Import of Python libraries and original data

In [263]:
# Let's import the necessary Python libraries
import pandas as pd
import numpy as np
from datetime import datetime
import datetime as dt

In [264]:
# Let's create a dataframe by importing the Linkfire's original data from the .csv file to which I saved the Google Sheets data
df = pd.read_csv('Linkfire.csv', sep=',')
df.name = 'linkfire'

<a id='03'></a>
#### 1.2. Review of the original table

In [265]:
# Let's see what we have...

# Let's see how random 10 rows of the table look
display(df.sample(10)) 

partner        date  visits  ad_impressions
22  partner2  2022-07-12   33305               0
17  partner2  2022-07-07   34330           17963
46  partner4  2022-07-08  555451          202404
40  partner3  2022-07-16       1               3
48  partner4  2022-07-10  537434          201600
26  partner2  2022-07-16   39021           19661
27  partner2  2022-07-17   40082           18575
66  partner5  2022-07-14  338768           64078
51  partner4  2022-07-13  520493               0
60  partner5  2022-07-08  413480           71612

In [266]:
# Let's see the main info about the table
display(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   partner         70 non-null     object
 1   date            70 non-null     object
 2   visits          70 non-null     int64 
 3   ad_impressions  70 non-null     int64 
dtypes: int64(2), object(2)
memory usage: 1.7+ KB


None

In [267]:
# Let's see the main statistics of the dataset
display(df.describe())

visits  ad_impressions
count      70.000000       70.000000
mean   199303.685714    57002.685714
std    208774.767478    70429.625060
min         1.000000        0.000000
25%     33307.500000        7.000000
50%     92154.500000    32377.000000
75%    388375.000000    63578.500000
max    583511.000000   218407.000000

In [268]:
# Let's see right away how many different partners we have
df['partner'].unique()

array(['partner1', 'partner2', 'partner3', 'partner4', 'partner5'],
      dtype=object)

We can immediately note the following: 

* The table has 70 records (rows) in total
* There are no blanks (no missing values), so we don't need to fill in the blanks
* The "date" column contains data of the type string (text) - we'd better change the format to datetime
* Table's columns have understandable names, so we don't need to rename them for clarity
* There are 5 different partners in total

[Up to the contents](#Contents)

<a id='04'></a>
### 2. Data pre-processing and optimization

<a id='05'></a>
#### 2.1. Check for full duplicates

In [269]:
# Let's show the sum of duplicated rows
df.duplicated().sum()

0

There are no duplicated rows, so we don't need to delete them. 

<a id='06'></a>
#### 2.2. Changing data formats for ease of use

In [270]:
# Let's change the string format of the column "date" to "datetime"
df['date'] = pd.to_datetime(df['date'])

In [271]:
# Let's control that the format change has been made, and the "date" column is now in datetime format
df.dtypes

partner                   object
date              datetime64[ns]
visits                     int64
ad_impressions             int64
dtype: object

Column values are now of the needed formats.

[Up to the contents](#Contents)

<a id='07'></a>
### 3. Exploratory data analysis

<a id='08'></a>
#### 3.1. Analysis of ad impressions 

In [272]:
# let's add a new column with weekday names for convenience
df['weekday'] = df['date'].dt.day_name()

# Let's put the new column next to the "date" column
df = df[['partner', 'date', 'weekday', 'visits', 'ad_impressions']]

# Let's see what we got
df.tail()

partner       date    weekday  visits  ad_impressions
65  partner5 2022-07-13  Wednesday  320541               0
66  partner5 2022-07-14   Thursday  338768           64078
67  partner5 2022-07-15     Friday  421557           68416
68  partner5 2022-07-16   Saturday  392375           55674
69  partner5 2022-07-17     Sunday  330021           47120

In [273]:
# Actually, since the dataframe is not very long - just 70 rows - I want to see it all and not just a selection. 
# Maybe my eyes will notice some regularities or patterns.

with pd.option_context('display.max_rows', None):
    print(df)

     partner       date    weekday  visits  ad_impressions
0   partner1 2022-07-04     Monday   90920           57702
1   partner1 2022-07-05    Tuesday   90622           59285
2   partner1 2022-07-06  Wednesday   91324           57795
3   partner1 2022-07-07   Thursday   92700           59718
4   partner1 2022-07-08     Friday   92011           61069
5   partner1 2022-07-09   Saturday   93633           62080
6   partner1 2022-07-10     Sunday   92298           61185
7   partner1 2022-07-11     Monday  110757           70354
8   partner1 2022-07-12    Tuesday   94363               0
9   partner1 2022-07-13  Wednesday   85600               0
10  partner1 2022-07-14   Thursday   73036           45063
11  partner1 2022-07-15     Friday   89742           55355
12  partner1 2022-07-16   Saturday  102304           60884
13  partner1 2022-07-17     Sunday   94211           58198
14  partner2 2022-07-04     Monday   34592           18744
15  partner2 2022-07-05    Tuesday   36458           186

Zeros for ad impressions seemingly always go in pairs. Let's only look at the rows with zero ad impressions - maybe we will notice some pattern

In [274]:
# Let's filter out the dataframe keeping only the rows with 0 as impressions
df.query('ad_impressions == 0')

partner       date    weekday  visits  ad_impressions
8   partner1 2022-07-12    Tuesday   94363               0
9   partner1 2022-07-13  Wednesday   85600               0
22  partner2 2022-07-12    Tuesday   33305               0
23  partner2 2022-07-13  Wednesday   31966               0
36  partner3 2022-07-12    Tuesday       2               0
37  partner3 2022-07-13  Wednesday       2               0
41  partner3 2022-07-17     Sunday       1               0
50  partner4 2022-07-12    Tuesday  547196               0
51  partner4 2022-07-13  Wednesday  520493               0
64  partner5 2022-07-12    Tuesday  328254               0
65  partner5 2022-07-13  Wednesday  320541               0

Ok, now it's clear that zero ad impressions only happened on two specific dates - **Tuesday 07.12.2022** and **Wednesday 07.13.2022**.

Also, we see that there were visits on those dates, so it's just the abcense of ad impressions. We will need to inquire why exactly these days ads were not shown to visitors - is it because the budget balance going down on those dates or for some technical reason, or maybe some experiment? This is the first question we should write down. 

Let's now analyze partners.

<a id='09'></a>
#### 3.2. Partner analysis 
It seems like all 5 partners have various daily audience and they also have different coverage of daily visits with ad impressions. Let's find out more. 

In [275]:
# let's first add a new column with percentage of daily coverage (how many visits "recieved" ad displays)
df['ad_coverage_%'] = ((df['ad_impressions'] / df['visits'])*100).round(2)
df.head(5)

partner       date    weekday  visits  ad_impressions  ad_coverage_%
0  partner1 2022-07-04     Monday   90920           57702          63.46
1  partner1 2022-07-05    Tuesday   90622           59285          65.42
2  partner1 2022-07-06  Wednesday   91324           57795          63.29
3  partner1 2022-07-07   Thursday   92700           59718          64.42
4  partner1 2022-07-08     Friday   92011           61069          66.37

In [276]:
# Let's now group by partners, display them in descending order by their daily median visits, and also see their daily
# median ad impressions and daily median ad coverage

mean_daily_visits_by_partners = df.groupby('partner').agg({'visits': 'median', 'ad_impressions':'median', 'ad_coverage_%':'median'})\
.sort_values(by = 'visits', ascending=False).reset_index()
mean_daily_visits_by_partners.columns = ['partner','median daily visits', 'median_daily_ad_impressions', 'median_daily_ad_coverage_%']
mean_daily_visits_by_partners.round(2)

partner  median daily visits  median_daily_ad_impressions  \
0  partner4             543501.5                     200432.0   
1  partner5             334394.5                      51397.0   
2  partner1              92154.5                      58741.5   
3  partner2              34335.5                      18039.5   
4  partner3                  3.0                          3.0   

   median_daily_ad_coverage_%  
0                       36.61  
1                       15.54  
2                       63.38  
3                       50.86  
4                       80.35

We can see that partner websites are really different in daily audience. While Partner 4 has about 543,000 daily visits (it's unclear to me whether those are unique visitors or all visits including repetitive visits of the same users), Partner 3 has about only 3 daily visits. 

I believe, we need to get more information about the partners and their specifics. Why Partner 3 has so few visits - is it an experimental or brand new website? Why is it then included in our marketing mix of partners? This is what we should investigate by asking colleagues. 

Another thing that we're missing is actually what are the budget spendings on each partner per ad impression? We don't know which ones are cheap and which are expensive. 

Finally, we don't have conversion data from impressions to clicks. Without cost per impression and ad conversion, we cannot really tell which platforms are more beneficial for budget spending. 

It only leaves us with the possibility to check the daily median "coverage" of ads (how many visitors are affected by ad displays). 

We see that different partners have different coverage - ad impressions to their visitors. For eaxample, for partner 1 ads are shown to about 63% of daily visitors (visits), and for partner 4 (who has the biggest daily audience) - the ads are shown on 36% of visits.

We should really ask more about the different partners and their specifics. 

[Up to the contents](#Contents)

<a id='10'></a>
### 4. Conclusions

Here's the summary of observations from the data set: 

* Partners are very different in terms of daily visits to their resources, **ranging from average of 3 visit for partner 3 to 534,500 visits for partner 4 per day**. We need to learn more about specifics of those partners and their audience. We should also ask **why do we need partner 3** with so few daily visits. 
* Partners are also different in terms of **how we spend the advertising budget on them (not in the money terms but in terms of ad impressions, because we don't know the costs per impression**). We need to further **inquire about costs per impression per partner**. That will give us the sum of ad costs per partner for that period of time, **so we know which partner is the most expensive and cheap for us**. 
* We **don't know the conversion rates from ad impressions to ad clicks** and transitions, so we deffinitely need to inquire about that to know how effectively budgets are spent on different partners. 
* On two specific dates - **Tuesday 07.12.2022 and Wednesday 07.13.2022** - there were no ad displays for all partners. Probably those days we were out of budget, or we stopped advertising, or there were technical issues or other reasons. We need to ask, maybe somebody can explain. 
